models/attachment.go

```diff
@@ -1,11 +1,151 @@
 package models
 
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+)
+
 // Attachment contains the fields and methods for
 // an email attachment
 type Attachment struct {
-	Id         int64  `json:"-"`
-	TemplateId int64  `json:"-"`
-	Content    string `json:"content"`
-	Type       string `json:"type"`
-	Name       string `json:"name"`
+	Id          int64  `json:"-"`
+	TemplateId  int64  `json:"-"`
+	Content     string `json:"content"`
+	Type        string `json:"type"`
+	Name        string `json:"name"`
+	vanillaFile bool   // Vanilla file has no template variables
+}
+
+// ValidateAttachment ensures that the provided attachment uses the supported template variables correctly.
+func (a Attachment) Validate() error {
+
+	vc := ValidationContext{
+		FromAddress: "foo@bar.com",
+		BaseURL:     "http://example.com",
+	}
+	td := Result{
+		BaseRecipient: BaseRecipient{
+			Email:     "foo@bar.com",
+			FirstName: "Foo",
+			LastName:  "Bar",
+			Position:  "Test",
+		},
+		RId: "123456",
+	}
+	ptx, err := NewPhishingTemplateContext(vc, td.BaseRecipient, td.RId)
+	if err != nil {
+		return err
+	}
+	_, err = a.ApplyTemplate(ptx)
+	return err
+}
+
+// ApplyTemplate parses different attachment files and applies the supplied phishing template.
+func (a *Attachment) ApplyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {
+
+	decodedAttachment := base64.NewDecoder(base64.StdEncoding, strings.NewReader(a.Content))
+
+	// If we've already determined there are no template variables in this attachment return it immediately
+	if a.vanillaFile == true {
+		return decodedAttachment, nil
+	} else {
+
+		// Decided to use the file extension rather than the content type, as there seems to be quite
+		//  a bit of variability with types. e.g sometimes a Word docx file would have:
+		//   "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
+		fileExtension := filepath.Ext(a.Name)
+
+		switch fileExtension {
+
+		case ".docx", ".docm", ".pptx", ".xlsx", ".xlsm":
+			// Most modern office formats are xml based and can be unarchived.
+			// .docm and .xlsm files are comprised of xml, and a binary blob for the macro code
+
+			// Zip archives require random access for reading, so it's hard to stream bytes. Solution seems to be to use a buffer.
+			// See https://stackoverflow.com/questions/16946978/how-to-unzip-io-readcloser
+			b := new(bytes.Buffer)
+			b.ReadFrom(decodedAttachment)
+			buf := b.Bytes()
+			zipReader, err := zip.NewReader(bytes.NewReader(buf), int64(len(buf))) // Create a new zip reader from the file
```

Comment:
I've been trying to figure this one out. Basically the problem is that zip archives require random access for reading, so it's hard to stream bytes (see [this post](https://stackoverflow.com/questions/16946978/how-to-unzip-io-readcloser)).

I did manage one piece of increased efficiency by getting the buffer length using `b.Len()`, and dropped the second buffer:

```
b := new(bytes.Buffer)
b.ReadFrom(decodedAttachment)
//buf := b.Bytes()
	
zipReader, err := zip.NewReader(bytes.NewReader(b.Bytes()), int64(b.Len())) 
```

But I can't find a way to use a solution like your proposed ```zipReader, err := zip.NewReader(b)```

I created a snippet on Go playground which decompresses a 'hello world' text file, in case that saves you time having a go: https://play.golang.org/p/g7VDSPXxvZU